text/0104-string-parameterization-output-spec.md

```diff
@@ -0,0 +1,264 @@
+
+Sentry performs parameterization in many different part of the application. The SDKs, Relay, and Ingest are three examples. There is no agreement between these parts beyond conventions. As a result, the output of parameterization is inconsistent.
+
+Here are three ways different SDKs might parameterize a SQL query:
+
+- `SELECT * FROM countries WHERE id = ?`
+- `SELECT * FROM countries WHERE id = %s`
+- `SELECT * FROM countries WHERE id = $1`
+
+As another example, the SDKs might pass the query `SELECT * FROM country WHERE code IN (?, ?)` to Ingest, which will further parameterize that as `SELECT * FROM country WHERE code IN (?)`. The string `(?, ?)` is technically only partially parameterized from the system's perspective.
+
+This has important downstream effects on user experience. A good parameterizer will correctly group similar strings together in a way that represents the original intent. If parameterization is successful, users will see groups that make sense to them. Good parameterization will also reduce cardinality, and therefore reduce data storage costs. A bad parameterizer will fill the users’s lists of transactions and spans with noisy similar descriptions that drown out the signal.
+
+This spec documents the desired output of a parameterizer. This makes it possible for different parameterizers to agree, conform to a standard, share code, and create consistency in the system.
+
+## Definitions
+
+- **Parameter**: A substring in a span description that represents a dynamic value that was substituted in by code. e.g., in the string `/users/17`, `17` is a likely parameter
+- **Parameterization**: The act of determining parameters, and replacing them with a substitution character. Also sometimes called "normalization"
+- **Parameterizer**: A piece of code, or a system that performs parameterization
+- **Substitution character**: A character or string of characters that denote a parameter. e.g., in the string `SELECT * FROM countries WHERE id = ?;` the `?` is a substitution character
+
+## Overall Guidelines
+
+- Parameterize *every* parameter. If a string contains multiple parameters, parameterize each one according to the relevant rules
+- Re-order unordered parameters in alphabetical order if it does not change the meaning of the string (e.g., see “Query Strings”)
+- Do not alter the string in any other way (e.g., do not re-format the whitespace, quotes, or any other symbols)
+
+## Parameterizing SQL Commands
+
+SQL queries usually appear in descriptions of `db` spans. In a SQL query (or any SQL statement) a parameter is any value that is either generated randomly, or supplied from a variable in the code. The following symbols are *not* considered parameters, and should not be altered or re-ordered:
+
+- column names
+- table names
+- aliases
+
+### Simple Parameters
+
+For individual parameters, **the parameterizer must replace the parameter with a single unquoted `?` character**.
```

Comment:
is the intent to get rid of all the variations like `%s` or `%(named_param)s`? Those templates are often 1:1 forwarded from the underlying DB driver, of which e.g. the Python SDK is mostly agnostic. Different DBs technically have different ways to substitute strings. They are very similar to each other as far as postgres/mysql/sqlite is concerned, but what about random non-SQL stuff like MongoDB queries?

Not clear to me where you want this normalization to happen. I think it would be good to only do it serverside in Relay, make some educated guesses on how string formatting works (similar to how we do it already for logentry) and preserve the original template somewhere in case the parametrizer misbehaves.

Reply:
> is the intent to get rid of all the variations like %s or %(named_param)s?

The intent is for the SDKs to use the same parameterization character, if that's what you mean!

> but what about random non-SQL stuff like MongoDB queries?

This is a good question, I can add a section below for parameterization structures queries like MongoDB (and Snuba, technically!)

> Not clear to me where you want this normalization to happen.

This spec have no opinion at all on _where_ parameterization happens. I mention this early in the description, but I should make that more explicit. We parameterize things in all kinds of places (like the SDKs, and serverside), to your point.

I do agree that preserving the original string makes a lot of sense, I cover that point a little further below in "Storing Parameterization Results"

Reply:
> This spec have no opinion at all on where parameterization happens.

@untitaker correct me if I'm wrong, but I think the concern is not about where the parameterization (i.e. replacing IDs with placeholders) happens, but where we ensure that these placeholders adhere to the standardization.

As @untitaker pointed out,

> Those templates are often 1:1 forwarded from the underlying DB driver, of which e.g. the Python SDK is mostly agnostic.

so we probably do not want to add the complexity of parsing and converting db-specific placeholders to standard-conform placeholders in the SDK itself.

Reply:
I'm curious to hear more about this, since converting db-specific placeholders to stands-conform is exactly what I'm suggesting! Is it a lot of work? Right now, the placeholders conform to the language of the SDK (or the framework) rather than the underlying database type. Unifying them would make downstream processing easier (see updated motivation section)

That said, I'm not suggesting we _have_ to unify them, but at a minimum I would like to document these placeholders, so there's a reference for which ones exist, and where, so any downstream processing can be consistent

Reply:
Unifying them is definitely possible, but I wonder whether users actually _want_ them unified. E.g. A react user might prefer seeing `/api/:user` over `/api/{user}`. This is ultimately a product decision though, I have no strong opinion.

An alternative way to make downstream processing aware of what parts of the string are placeholders is to store annotations outside of the string itself, e.g. by using the `_meta` protocol mentioned further below.